### DISCUSSION TITLE
Challenging typescript errors

### DISCUSSION THREAD

**Discussion by remy90** [react-hook-form/react-hook-form#8175] [Q&A]
I'm trying to address a typescript error relating to a [Smart form component](https://www.react-hook-form.com/advanced-usage#SmartFormComponent) _with a Controller wrapper_. 

To summarise the code below: if the component passed into <Form> has prop 'InputName', wrap it in a Controller. Does anyone have any suggestions?

sandbox: https://codesandbox.io/s/dreamy-edison-s8nycy?file=/src/Form.tsx

```
interface IFormFieldProps {
  inputName: string;
  label?: string | JSX.Element;
  field?: any;
  fieldState?: ControllerFieldState;
  formState?: any;
  disabled?: boolean;
  showErrorOnDirty?: boolean;
}

export const Form = ({ defaultValues, children, onSubmit, formSchema }: IFormTemplateProps) => {
  const methods = useForm<IFormFieldProps>({
    resolver: yupResolver(formSchema),                            // Error 1
  });

  return (
    <FormProvider {...methods}>
      <form onSubmit={methods.handleSubmit(onSubmit)}>
        { React.Children.map(children, (child: ReactElement<IFormFieldProps>) =>
          !child?.props?.inputName
            ? child
            : (
            <Controller                                           // Error 2
              name={child.props.inputName}
              control={methods.control}
              render={({ field, fieldState, formState }) => (
                  React.createElement(child.type, {
                    ...{
                      ...child.props,
                      key: child.props.inputName,
                      field, fieldState, formState, 
                    }
                })
              )}
            />)
          )
        }
        <ButtonContainer reverse>
          <Button onClick={() => reset()}>Cancel</Button>
          <Button type="submit">Save</Button>
        </ButtonContainer>
      </form>
    </FormProvider>
  );
}
```

Error 0: If <IFormFieldProps> is removed from useForm; `Type 'string' is not assignable to type 'never'.ts(2322)`

Error 1: `Type '<TFieldValues extends FieldValues, TContext>(values: UnpackNestedValue<TFieldValues>, context: TContext | undefined, options: ResolverOptions<TFieldValues>) => Promise<...>' is not assignable to type 'Resolver<IFormFieldProps, any>'.` (amongst others that follow)

Error 2: `Type instantiation is excessively deep and possibly infinite.ts(2589)
Type of property 'all' circularly references itself in mapped type '{ [K in keyof Document]-?: PathImpl<K & string, Document[K]>; }'.ts(2615)`. Pretty sure its not infinite, not sure why instantiation is excessively deep.

**Comment by Moshyfawn** [comment]
It's insanely hard to debug a TS issue with only a plain un-formatted code provided.

Please, consider creating a minimal reproduction codesanbox instead.

**Comment by remy90** [comment]
Hi @Moshyfawn 
please find a minimal repro here: https://codesandbox.io/s/dreamy-edison-s8nycy?file=/src/Form.tsx

**Comment by Moshyfawn** [comment]
You're passing your `IFormFieldProps` interface into the `useForm` generic
```ts
useForm<IFormFieldProps>()
```

The `IFormFieldProps` interface contains some custom complex types like `JSX.Element` and an actual RHF type `ControllerFieldState`. RHF supports only plain objects, non-flat arrays, `string`, `number` etc. Hence the issue.

Consider removing / simplifying the custom objects from your form and providing the form interfaces accordingly.
